ldap/ad web ui proposal (#1116)
vmware-tanzu/pinniped: proposals/1113_ldap-ad-web-ui/README.md

* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)
* Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
* Prevent LDAP injection attacks
* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
* Screens are accessible and friendly to screen readers
* Screens are friendly to password managers

#### Non-goals
* A rich client (ie the use of javascript)
* Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.
* Branding & customization beyond the information listed in the goals used to identify the login screen belongs to Pinniped.
* Supporting SSO integrations
* Internationalization or localization. The CLI doesn't currently support this either.

### Specification / How it Solves the Use Cases

#### API Changes

The supervisor must accept requests from other clients, as detailed
in the (todo) proposal for dynamic client registration.

> When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their

Reviewer: I should be able to use the `pinniped-cli` OAuth client in this way as well.  For example, if this client is used and the username/password headers are not set, it should redirect to the login page.  We need to be able to test this flow without adding a new client.

IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
When a client other than the Pinniped CLI makes an authorization endpoint request with
custom Username/Password headers, they should be rejected.

The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable or when

> a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in

Reviewer: I am unsure if we need a CLI flag for this.

Reviewer 2: Based on https://github.com/vmware-tanzu/pinniped/pull/1116#discussion_r850455917 which says:

> Newer CLIs should fallback to this option when there is no TTY/env vars set. I think more of the flow logic needs to be moved into the login command and out of the get kubeconfig command (to allow this decision to be made at runtime).

I think you're saying that the CLI can make the decision to direct the user to the web UI based on data available at runtime.  Is that right?

Reviewer: > I think you're saying that the CLI can make the decision to direct the user to the web UI based on data available at runtime. Is that right?

Yeah, and that the `login oidc` command already has over 20 flags, and I would like to not add more unless we really need to.

Author: Resolving this because the existing `--upstream-identity-provider-flow` flag covers this case.

Author: Resolved because the existing `upstream-identity-provider-flow` flag covers it

> a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in

Reviewer: I am unsure if we need a CLI flag for this.

instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
may wish to authenticate using the pinniped cli but without access to a terminal.

#### Upgrades

This change is backwards compatible. Users would see no changes unless they decided to register

> a new client or change the pinniped cli flags.

Reviewer: I would expect the discovery metadata to change to state that the LDAP/AD IDPs also support `browser_authcode` now.  Newer CLIs should fallback to this option when there is no TTY/env vars set.  I think more of the flow logic needs to be moved into the `login` command and out of the `get kubeconfig` command (to allow this decision to be made at runtime).

Reviewer 3: > I would expect the discovery metadata to change to state that the LDAP/AD IDPs also support browser_authcode now.

Yes. The `pinniped get kubeconfig` command already supports this by reading the results of the IDP discovery endpoint and allowing whatever the results of that endpoint dictate. So kubeconfigs generated by both old and new CLIs would be the same.

An old CLI will output a friendly validation error during login if the kubeconfig chooses `browser_authcode` for an LDAP/AD IDP. This would be solved by upgrading the CLI to a version which skips that validation. That seems like reasonable upgrade behavior. It could be documented in the release notes.

> I think more of the flow logic needs to be moved into the login command and out of the get kubeconfig command (to allow this decision to be made at runtime).

The runtime decision to prefer the `browser_authcode` flow over the `cli_password` flow could be done with a new environment variable which forces the use of `browser_authcode` regardless of the setting of `--upstream-identity-provider-flow` from the kubeconfig. That would be sufficient to handle use cases like IDE integrations which don't have access to a terminal to prompt for usernames/passwords and need to override the flow selection. That could be part of this proposal, or could be a separate small feature after this proposal is implemented.

In the future, we could offer more capabilities to optionally move the selection of IDP and flow into the end user's runtime choice during login, however that's probably beyond the scope of this proposal. For the purposes of this proposal, the CLI's existing implementation should be good enough.

Reviewer 3: See also https://github.com/vmware-tanzu/pinniped/pull/1132#pullrequestreview-948919721 regarding the behavior of `pinniped get kubeconfig`.

Reviewer 3: I think this conversation was resolved in offline discussions and the doc was updated accordingly.

However if they do choose to register a new client they may need to update the following:
- FederationDomains today may be using private certificate authorities. These are trusted
  for our use case but a browser will flag them as unsafe. Admins will have to transition to letsencrypt
  or another public Certificate Authority to prevent making end users click past messages about the certificate
  being untrusted.
- The name of the idp custom resource is currently not published to users logging in with Pinniped.
  We plan on exposing this to indicate to users which idp they are logging in to.

>   Admins may need to update this to something more user-friendly.

Reviewer: It is part of our custom discovery metadata IIRC.

Author: It's always been available but not advertised. Now it will be advertised, which seemed worth calling out.

Reviewer: Do we really need to expose the name?  You can only have one IDP configured, so if the IDP name did not matter for the CLI, why does it matter in the UI?

Author: Honestly I was wondering the same thing, the document that Jeremy and Anjali came up with had it as one of their goals which is why I put it here as one. 
Perhaps we are expecting admins to have communicated instructions about what IDP they're logging in to when they distribute the admin kubeconfig? CLI users may be primed about how to use Pinniped in a way that users of dashboards might not be.

Reviewer 2: Some things to consider:
* This proposal will result in a minimally branded login page. That branding will almost definitely not match enterprise branding.  While it isn’t much, providing the IdP name offers a small clue the user that they're in the right place and not some place trying to steal their password.
* Multi IdP will eventually arrive, might as well start prepping.
* I imagine the CLI being brought in line to show the IdP name too.

Reviewer 3: > Admins may need to update this to something more user-friendly.

This seems like a good idea. Without it, we could only assume that we should use the name of the IDP CR, which is not allowed to be formatted very well (e.g. no spaces). I could imagine adding this as a new optional field in the OIDC/LDAP/ActiveDirectoryIdentityProvider CRDs. When set, the name could be shown in the new LDAP/AD login UI, and shown by the CLI above the login prompts for LDAP/AD and OIDC password grant. When not set, those places could show the name of the IDP CR.

I could also imagine adding a similar field in the same places to allow the admin to rename the username prompt (e.g. name it "Email Address"). This may be outside the scope of this proposal, but we could leave room for it if we add a new API field.

All of these optionally configured strings could be added to the IDP discovery result to be consumed by `pinniped get kubeconfig` and encoded into the kubeconfig as CLI flags.

Maybe allowing configuring a login banner message could fit into the same place someday too.

Author: Resolved, see next sentence about leaving further branding decisions until later.

>   Admins may need to update this to something more user-friendly.

Reviewer: It is part of our custom discovery metadata IIRC.

#### Tests

Chromedriver browser based integration tests will be needed to ensure that a user can log in from a web-based app

> by entering their ldap credentials into the web page.

Reviewer: Outline the high level scenarios we would test such as CSRF protection, valid user/pass, invalid user/pass, etc

> by entering their ldap credentials into the web page.

Reviewer: Outline the high level scenarios we would test such as CSRF protection, valid user/pass, invalid user/pass, etc

#### New Dependencies

> This should be kept to a very simple HTML page with minimal, clean CSS styling.

Reviewer: Will this page mention Pinniped in any way?

Author: Yep, called out in the goals section.

Reviewer 3: It would make sense for the default style of this page to match the Supervisor's existing web page (the page that shows authcodes). See html and css in this directory: https://github.com/vmware-tanzu/pinniped/tree/fffcb7f5b4558b102ca1065a9c7eec2564577396/internal/oidc/provider/formposthtml

Note that the existing page has a tiny little Pinniped favicon for those browsers which support favicons on browser tabs, but otherwise does not mention "Pinniped" or show a Pinniped logo. This was done intentionally at the time of the development of that page to make it branding-agnostic until a feature could be added for branding configuration support. We might want to consider that approach again.

> This should be kept to a very simple HTML page with minimal, clean CSS styling.

Reviewer: Will this page mention Pinniped in any way?

Javascript should be avoided.

#### Observability Considerations
* Logging login attempts at higher log levels.

#### Security Considerations
* Preventing LDAP injection attacks: this should be done server-side using our existing
  string escaping.
* CSRF protection via a CSRF cookie: this should be similar to the way it is done for the
  OIDCIdentityProvider today
* The new UI page must be HTTPS.

#### Documentation Considerations
This new feature will require documentation to explain how to configure it and to publicise that it is available.
This should include:
* A blog post describing the feature
* Website documentation in the form of a how-to guide

### Other Approaches Considered
Today, users can configure Dex if they want a web-based LDAP login.
This introduces complexity because they have to install, configure and
maintain both Pinniped and Dex in order to use this feature. It also means
that users do not benefit from the opinionated `ActiveDirectoryIdentityProvider`
config because Dex does not have an equivalent.

## Open Questions

> * Is the ldap login page hosted per FederationDomain?

Reviewer: It does not really matter today because you can only have one IDP.

Reviewer 2: How does the CLI handle this today? The CLI OAuth client isn't associated with a federation domain.  What about the APIs that orchestrate the upstream login, are they associated with a federation domain?

Reviewer: The CLI does not really know anything about a federation domain - it just looks at the issuer URL to perform discovery to find the generic OIDC endpoints and the pinniped specific endpoints.  In that sense, all end-user facing supervisor APIs are associated with a federation domain because there simply is not any other place to put them.

Author: I removed this as an open question because I agree with Mo that it does not matter much today as we have one IDP, and the question felt too broad.
At the moment I think there should be one login page hosted at `issuer-url/some/path`, which in practice means it is associated with a FederationDomain as that's where the issuer is defined. Does that make sense to folks or do we still need to talk through this more?

Reviewer 3: I think that makes sense. By putting it on the FederationDomain's endpoints, then the TLS serving certs will be handled correctly by our existing configuration options and implementation code. Note that TLS serving certs are already per-FederationDomain.

> * Is the ldap login page hosted per FederationDomain?

Reviewer: It does not really matter today because you can only have one IDP.

* What is the format for the URL? (`issuer/some/path`? Something else?) Can we make it so we can reuse the existing cert,

>   or will we need a new wildcard cert?

Reviewer: Requiring a new cert will make this excessively painful for admins.

Author: `<issuer_path>/ldap/<ldap_idp_name>/login` was suggested in https://hackmd.io/@pinniped/Hk2WA1sb_#Possible-Future-Feature-Web-Based-LDAP-Login-Support, I think that makes sense and would not require new certs.

Reviewer: This URL is not an API right?  i.e. we redirect to it, and we can change that at any point (the starting point is always the authorize endpoint).  Thus I think `<issuer_path>/login` is sufficient?

Reviewer 3: Putting the IDP name and type in the path was a suggestion to help get ready to handle the future feature of multiple IDP support. The information does not necessarily need to be in the path if there is some other way for the authorize endpoint to communicate the requested IDP name and type through the redirect to the handler for the LDAP/AD login page. The login page will need to be able to use the name/type to lookup the IDP, for example to show its name on the page.

Reviewer 3: Related thought: If the IDP name/type is encoded into a signed/encrypted state param (rather than the path), then we can rest assured that nobody can try to directly access this endpoint without going through the authorize endpoint first. That would help us avoid needing to gracefully handle error edge cases like someone directly trying to access the new login endpoint to log into an OIDCIdentityProvider (not supported), since that would be effectively impossible because the authorize endpoint would never send them to the login endpoint for an OIDCIdentityProvider.

Reviewer 3: I think this conversation was resolved and the doc was updated accordingly.

>   or will we need a new wildcard cert?

Reviewer: Requiring a new cert will make this excessively painful for admins.

However, Pinniped is planning to introduce support for dynamic OAuth clients.
These clients should _not_ be trusted to handle a user's IDP credentials.
Therefore, we need a mechanism for untrusted clients to acquire Pinniped's downstream tokens while
leaving the IDP credential handling to the Pinniped supervisor.

## Proposal
Pinniped must provide a simple login screen in order to support UIs that wish
to authenticate with the Pinniped Supervisor to gain access to a cluster without
requiring each app to handle IDP credentials.

### Goals and Non-goals

#### Goals
* Prevent OAuth clients, other than the Pinniped CLI, from providing credentials via the authorization request
* Provide a minimal feature set (ie user id, password & submit button only)
* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
  but do not expose whether it's the username or password that's incorrect)
* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)

> * Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)

Reviewer 3: Tiny typo: meant to say "web forms"

* Prevent LDAP injection attacks
* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
* Screens are accessible and friendly to screen readers
* Screens are friendly to password managers

#### Non-goals
* A rich client (ie the use of javascript)

> * Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.

Reviewer 3: Small nit: We can't leave these particular features to the actual identity provider to implement, since this page will be inside the Supervisor's code. However, the point about these being non-goals for the first version of this page still stands.

* Supporting SSO integrations
* Internationalization or localization. The CLI doesn't currently support this either.

### Specification / How it Solves the Use Cases

#### API Changes

The supervisor must accept requests from other clients, as detailed
in the (todo) proposal for dynamic client registration.
When a client other than pinniped-cli makes an authorization endpoint request with `response_type=code` and their
IDP is an LDAP or Active Directory IDP, the user will be redirected to the new login page.
The login page should display the IDP name and indicate that it belongs to Pinniped.
When a client other than the Pinniped CLI makes an authorization endpoint request with
custom Username/Password headers, they should be rejected.

The discovery metadata for LDAP/AD IDPS should indicate that they support a flow of `browser_authcode`.

The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable,
it will open a browser to log in
instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)

> may wish to authenticate using the pinniped cli but without access to a terminal.

Reviewer 3: Rather than making decision based on tty availability, the CLI should use the existing `--upstream-identity-provider-flow` flag. This has already been implemented in both `pinniped login oidc` and `pinniped get kubeconfig`. More details in thread on k8s Slack: https://kubernetes.slack.com/archives/C01BW364RJA/p1650306152560689

The existing tty available/unavailable in the CLI behavior should not need to change. It's there to correctly handle cases like where the user is using kubectl in a shell script or where the user is piping input into kubectl. We wouldn't want the browser to open in those cases.

`UpstreamStataParamData.FormatVersion`, which would mean that logins in progress at the time of
upgrade would fail.

The pinniped cli should default to using the cli-based password flow, but when the `--upstream-identity-provider-flow`,
flag specifies `browser_authcode`, it will open a browser to log in
instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
may wish to authenticate using the pinniped cli but without access to a terminal.

Here is how the login flow might work:
1. The supervisor receives an authorization request.
   1. If the client_id param is not "pinniped-cli", and it includes username and password via the custom headers, reject the request.
   2. If the request does not include the custom username/password headers, assume we want to use the webpage login.
   3. Some day we may want to allow specifying the idp name and type as request parameters, but
      for now we do not have multiple idps.
   4. Encode the request parameters into a state param like is done today for the `OIDCIdentityProvider`.
      In addition to the values encoded today (auth params, upstream IDP name, nonce, csrf token and pkce),

>       encode the upstream IDP type.

Reviewer: This does mean that one cannot bookmark the `login` page.  Not sure if we care about that.

Reviewer 3: That's a good point. At the moment, the login page is only an interstitial in a downstream OIDC authcode flow. It doesn't have any meaning outside of that context, because there is currently no concept of starting a Supervisor session outside of a particular OIDC authorization invocation (which decides the client, the scopes, and all the other context of the session). So today there would be no reason to allow an end user to bookmark the login page. Maybe that could have some meaning in the future, at which time we could offer support for a different way to invoke the `GET /login` endpoint without a `state` param.

Reviewer 3: I think this conversation was resolved.

However, Pinniped is planning to introduce support for dynamic OAuth clients.
These clients should _not_ be trusted to handle a user's IDP credentials.
Therefore, we need a mechanism for untrusted clients to acquire Pinniped's downstream tokens while
leaving the IDP credential handling to the Pinniped supervisor.

## Proposal
Pinniped must provide a simple login screen in order to support UIs that wish
to authenticate with the Pinniped Supervisor to gain access to a cluster without
requiring each app to handle IDP credentials.

### Goals and Non-goals

#### Goals
* Prevent OAuth clients, other than the Pinniped CLI, from providing credentials via the authorization request
* Provide a minimal feature set (ie user id, password & submit button only)
* Provide generalized error messaging for failed logins that do not expose sensitive information (i.e. we should say "invalid username or password"
  but do not expose whether it's the username or password that's incorrect)
* Provide information easily allowing a user to identify the screen as belonging to Pinniped and which upstream IdP is being represented (e.g. IdP name)

> * Address basic security concerns for web firms (HTTPS, passwords use a password field, CSRF protection, redirect protection)

Reviewer 3: ```suggestion
* Address basic security concerns for web forms (HTTPS, passwords use a password field, CSRF protection, redirect protection)
```

* Prevent LDAP injection attacks
* Rely on the upstream IdP to address advanced security concerns (brute force protection, username enumeration, etc)
* Screens are accessible and friendly to screen readers
* Screens are friendly to password managers

#### Non-goals
* A rich client (ie the use of javascript)

> * Advanced UI features (e.g. remember me, reveal password). These features are better left to identity providers to implement.

Reviewer 3: ```suggestion
* Advanced UI features (e.g. remember me, reveal password)
```

The state param should be augmented to include the IDP type as well as the IDP name. The type
should be included in `UpstreamStateParamData` so that later when we get it back in the callback
request we can tell which IDP it is referring to. This will require an update to
`UpstreamStataParamData.FormatVersion`, which would mean that logins in progress at the time of
upgrade would fail.

The pinniped cli should default to using the cli-based password flow, but when the `--upstream-identity-provider-flow`,
flag specifies `browser_authcode`, it will open a browser to log in
instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
may wish to authenticate using the pinniped cli but without access to a terminal.

Here is how the login flow might work:
1. The supervisor receives an authorization request.
   1. If the client_id param is not "pinniped-cli", and it includes username and password via the custom headers, reject the request.
   2. If the request does not include the custom username/password headers, assume we want to use the webpage login.

>    3. Some day we may want to allow specifying the idp name and type as request parameters, but
>       for now we do not have multiple idps.

Reviewer 3: ```suggestion
   3. Today the CLI specifies the IDP name and type as request parameters, but the server currently ignores these
      since the Supervisor does not allow multiple idps today. This could be enhanced in the future to use the requested IDP when the params are present, and to show another UI page to allow the end user to choose which IDP when the params are not present. This leaves room for future multiple IDP support in this flow, however, the details are outside the scope of this proposal.
```

  is now displayed.

To enable users to upgrade smoothly, the behavior of the Pinniped CLI when it encounters multiple possible flow options will change.
Previously, the team had decided that the CLI should fail when there were multiple options (e.g. when it's could
use either the `browser_authcode` flow or the `cli_password` flow). However, that behavior would break existing
kubeconfigs once the `browser_authcode` flow was introduced to the IDP discovery doc.
Instead we are opting to prioritize based on the order listed in the IDP discovery doc.
Users will still have the option to override this priority with the `--upstream-identity-provider-flow` flag,
but that flag will not be required.

#### Tests

Chromedriver browser based integration tests will be needed to ensure that a user can log in from a web-based app
by entering their ldap credentials into the web page, as well as unit tests.

With the pinniped cli:
- succeeds with correct username and password
- fails with incorrect username, shows useful but nonspecific error message
- fails with incorrect password, shows useful but nonspecific error message
- with tty access, prompts for username and password on the cli

> - without tty access, opens a browser
> - without tty access, if the form post fails, don't ask user to copy and paste the authcode (we already know you have no tty to paste it into...)

Reviewer 3: ```suggestion
```

> - with tty access, prompts for username and password on the cli
> - without tty access, opens a browser
> - without tty access, if the form post fails, don't ask user to copy and paste the authcode (we already know you have no tty to paste it into...)

Reviewer 3: ```suggestion
```

Once dynamic clients are implemented:
- fails when attempting to pass username/password as headers on requests to the authorize endpoint
- tests of the rest of the dynamic client functionality that should be detailed as part of that proposal

#### New Dependencies
This should be kept to a very simple HTML page with minimal, clean CSS styling.
Javascript should be avoided.
The styling should match the [form post html page](https://github.com/vmware-tanzu/pinniped/tree/main/internal/oidc/provider/formposthtml)
as much as possible, we should reuse some of the existing css and add to it to keep the style consistent.

#### Observability Considerations

> * Logging login attempts at higher log levels.

Reviewer 3: @margocrawf What did you mean by this? Is this an audit logging concern, or something else?

Author: Yes, I was thinking some basic logging of the requests we're making. `upstreamldap.go` already has this. I'll update that to say something along those lines.

  string escaping.
* CSRF protection via a CSRF cookie: this should be similar to the way it is done for the
  OIDCIdentityProvider today
* The new UI page must be HTTPS.

#### Documentation Considerations
This new feature will require documentation to explain how to configure it and to publicise that it is available.
This should include:
* A blog post describing the feature
* Website documentation in the form of a how-to guide

### Other Approaches Considered
Today, users can configure Dex if they want a web-based LDAP login.
This introduces complexity because they have to install, configure and
maintain both Pinniped and Dex in order to use this feature. It also means
that users do not benefit from the opinionated `ActiveDirectoryIdentityProvider`
config because Dex does not have an equivalent.

## Open Questions
* What is the format for the URL? (`issuer/some/path`? Something else?) Can we make it so we can reuse the existing cert,
  or will we need a new wildcard cert?
* Currently we have little validation on branding requirements. Is specifying the idp name enough for users to understand
  how to log in?

> * How many users will be blocked on using this feature until they can have a company name and logo on the login page?

Reviewer 3: Can we move these to the "Answered Questions" section and write down our current answers?

## Implementation Plan
While this work is intended to supplement the dynamic client work, parts of it
can be implemented independently.

> The pinniped cli can support a web based ui flow via a command line flag, environment variable or checking whether a tty is available.

Reviewer 3: ```suggestion
The pinniped cli can support a web based ui flow via a command line flag or environment variable.
```

---
title: "Web UI for LDAP/AD login"
authors: [ "@margocrawf" ]

> status: "draft"
> approval_date: ""

Reviewer 3: ```suggestion
status: "accepted"
approval_date: "May 11, 2022"
```

  string escaping.
* CSRF protection via a CSRF cookie: this should be similar to the way it is done for the
  OIDCIdentityProvider today
* The new UI page must be HTTPS.

#### Documentation Considerations
This new feature will require documentation to explain how to configure it and to publicise that it is available.
This should include:
* A blog post describing the feature
* Website documentation in the form of a how-to guide

### Other Approaches Considered
Today, users can configure Dex if they want a web-based LDAP login.
This introduces complexity because they have to install, configure and
maintain both Pinniped and Dex in order to use this feature. It also means
that users do not benefit from the opinionated `ActiveDirectoryIdentityProvider`
config because Dex does not have an equivalent.

## Answered Questions
* Q: What is the format for the URL? (`issuer/some/path`? Something else?)
  A: `<issuer>/login`
* Q: Can we make it so we can reuse the existing cert, or will we need a new wildcard cert?
  A: Since the page is hosted on the issuer, we can reuse the existing `FederationDomain` cert.

> 
> ## Open Questions

Reviewer 3: ```suggestion
```

> 
> ## Open Questions
> * Q: Currently we have little validation on branding requirements. Is specifying the IDP name enough for users to understand
>   how to log in? How many users will be blocked on using this feature until they can have a company name and logo on the login page?
>   A: For our initial release, we will only specify the IDP name. We are open to adding further customization in response to feedback
>   from users once the feature is released.

Reviewer 3: ```suggestion
* Q: Currently we have little validation on branding requirements. Is specifying the IDP name enough for users to understand
  how to log in? How many users will be blocked on using this feature until they can have a company name and logo on the login page?
  A: For our initial release, we will only specify the IDP name. We are open to adding further customization in response to feedback
  from users once the feature is released.

## Open Questions
None.
```

The state param should be augmented to include the IDP type as well as the IDP name. The type
should be included in `UpstreamStateParamData` so that later when we get it back in the callback
request we can tell which IDP it is referring to. This will require an update to
`UpstreamStataParamData.FormatVersion`, which would mean that logins in progress at the time of
upgrade would fail.

The pinniped cli should default to using the cli-based password flow, but when the `--upstream-identity-provider-flow`,
flag specifies `browser_authcode`, it will open a browser to log in
instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
may wish to authenticate using the pinniped cli but without access to a terminal.

Here is how the login flow might work:
1. The supervisor receives an authorization request.
   1. If the client_id param is not "pinniped-cli", and it includes username and password via the custom headers, reject the request.
   2. If the request does not include the custom username/password headers, assume we want to use the webpage login.
   3. Some day we may want to allow specifying the idp name and type as request parameters, but

>       for now we do not have multiple idps.

Reviewer 3: Note that the CLI will already send these as parameters, but the server-side code for the authorization endpoint currently ignores them, as you pointed out, since we don't support multiple IDPs yet. Part of implementing multiple IDP support will involve paying attention to these params.

logo and the IDP name.
   1. The submission should be POST `<issuer-url>/login`.
   2. The state param’s value is written into a hidden form input, properly escaped.
   3. Username and password form inputs are shown.
4. The supervisor receives the POST request.
   1. Decode your state form param to reconstitute the original authorization request params
   (the client’s nonce and PKCE, requested scopes, etc) and also compare the incoming CSRF cookie to the value
   from the state param. This code would be identical to what we do in the upstream OIDC callback endpoint today.
   If the decoded state param’s timestamp is too old, it might be prudent to reject the request.
   2. Using the idp name/type from the state param, look up the IDP, bind to it, verify the username/password and
   get the users downstream username and groups.
   3. If the login succeeds, mint an authcode and store the session as a secret the same way as we do on the
   callback endpoint today. Redirect to the client’s redirect URI with the new authcode.
   4. If the login fails, respond with an error so the login page can render an error message. Allow the user to retry

>    login the same way we do with the CLI today (we leave brute force protection to the IDP)

Reviewer 3: I wonder if the POST handler should return a redirect to the GET handler to render the login page again? That might be more browser-friendly in regard to things like the behavior of the browser's "back" and "refresh" buttons. E.g. refresh would GET the login page again instead of posting the web form again. This would imply that the POST handler would have to inform the GET handler that it should render the error message onto the login page via query param or cookie (probably query param).
